### ISSUE TITLE
Parser: Support User-Defined Particle Attributes

### ISSUE THREAD

**Issue by ax3l** [ECP-WarpX/WarpX#3051]
For our filters and histograms in I/O (1D, 2D, full, #3033), it would be very useful if user-defined particle attributes could be used in parser functions.

Currently, we only support the pre-defined attributes there, most of the time that's `x,y,z,ux,uy,uz`.

Do you have an idea how we could do this @WeiqunZhang @RevathiJambunathan @hklion ?

**Comment by hklion** [comment]
Maybe a way to do this is to add a compile flag that controls the max number of user-defined attributes that can be used in the parsers? That way it can scale to something arbitrarily large if a user wants, but by default we don't have to carry around much (or any!) extra memory. Identifying which variables to use from the parser function would require some string gymnastics, but I don't think there's any reason variable names would have to be known at compile time (though I've never tried it).

**Comment by kale-j** [comment]
Are there currently any plans to finish this up and merge it in? 

I have a problem that this functionality would help for (one can potentially use attributes to track a randomly chosen subset of the particles, including with < 1 particle/cell). My particular use case could also be done with a standalone attribute filter (i.e., include the particle only if the given attribute is nonzero), which is what I will code up if there are no plans to revive this. This unfortunately wouldn't address the parser function flexibility problem, which is perhaps a bit more complex that I have time to dive into at the moment.

**Comment by hklion** [comment]
I don't have a generalized version that can be merged into development, but I do have a hacky version in a branch, see e.g. https://github.com/RevathiJambunathan/WarpX/blob/HarrisSheetinX/Tools/ReconnectionInputAndTools/inputs_reconnection_s30_attr.2d for how it gets used. The added user attribute is called `us` here. If this kind of interface would be of interest, I can provide more details about what I did.

I made an attempt at generalizing all of this a while back (see #3086), but never had the time to see it through myself. I think it's unlikely that I will in the foreseeable future.

**Comment by kale-j** [comment]
Yes, I am interested. It's okay if it's kind of hacky as I have enough of an idea of what to expect that I should be able to tell if anything is broken. My immediate use case is to use a particle attribute in plot_filter_function when outputting particle data. It looks like what you have coded may fit this, though I didn't see "us" defined anywhere in the input deck -- did you hard code this in the source itself?

A potential complication to note is that I may need to merge your changes into the latest version to use some newer features, but hopefully this will not be too bad. Is #3086 the best branch to use?
